File: partiql-plan/src/main/java/org/partiql/plan/rex/RexBag.java
package org.partiql.plan.rex;

import org.jetbrains.annotations.NotNull;
import org.partiql.plan.Operand;
import org.partiql.plan.Operator;
import org.partiql.plan.OperatorVisitor;
import org.partiql.spi.types.PType;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * Logical bag expression abstract base class.
 */
public abstract class RexBag extends RexBase {

    /**
     * @return new RexBag instance
     */
    @NotNull
    public static RexBag create(@NotNull Collection<Rex> values) {
        return new Impl(values);
    }

    /**
     * @return the values of the bag, also the operands (unordered).
     */
    @NotNull
    public abstract Collection<Rex> getValues();

    @NotNull
    @Override
    protected final RexType type() {
        return RexType.of(PType.bag());
    }

    @NotNull
    @Override
    protected final List<Operand> operands() {
        // uh oh! prescribing order (??)
        List<Operator> values = new ArrayList<>(getValues());
        Operand c0 = Operand.vararg(values);
        return List.of(c0);
    }

    @Override
    public <R, C> R accept(@NotNull OperatorVisitor<R, C> visitor, C ctx) {
        return visitor.visitBag(this, ctx);
    }

    private static class Impl extends RexBag {

        private final Collection<Rex> values;

        private Impl(Collection<Rex> values) {
            this.values = values;
        }

        @NotNull
        @Override
        public Collection<Rex> getValues() {
            return values;
        }
    }
}

File: partiql-plan/src/main/kotlin/org/partiql/plan/v1/operator/rex/RexBag.kt
package org.partiql.plan.v1.operator.rex

import org.partiql.types.PType

/**
 * TODO DOCUMENTATION
 */
public interface RexBag : Rex {

    public fun getValues(): Collection<Rex>

    override fun getChildren(): Collection<Rex> = getValues()

    override fun <R, C> accept(visitor: RexVisitor<R, C>, ctx: C): R = visitor.visitBag(this, ctx)
}

/**
 * Default [RexBag] operator for extension.
 */
internal class RexBagImpl(values: Collection<Rex>) : RexBag {

    // DO NOT USE FINAL
    private var _values = values

    override fun getValues(): Collection<Rex> = _values

    override fun getChildren(): Collection<Rex> = _values

    override fun getType(): PType = PType.bag()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is RexBag) return false
        if (_values != other.getValues()) return false
        return true
    }

    override fun hashCode(): Int = _values.hashCode()
}
